client/my-sites/stats/stats-module/index.jsx

```diff
@@ -70,6 +93,75 @@ class StatsModule extends Component {
 			// eslint-disable-next-line react/no-did-update-set-state
 			this.setState( { loaded: false } );
 		}
+
+		const isRealTime = this.props.query?.interval !== undefined;
+		if ( ! isRealTime ) {
+			return;
+		}
+
+		// Limit data processing to match the query interval to avoid spurious updates.
+		// Need to convert from milliseconds to seconds for comparison.
+		const limit = this.props.query.interval / 1000;
+		const { lastUpdated } = this.state;
+		const now = moment();
+
+		if ( ! lastUpdated || now.diff( lastUpdated, 'seconds' ) >= limit ) {
+			const updatedHistory = this.updateHistory( this.state.dataHistory, this.props.data );
+			const firstSnapshot = updatedHistory[ 0 ];
+			const lastSnapshot = updatedHistory[ updatedHistory.length - 1 ];
+			const diffData = this.calculateDiff( firstSnapshot.data, lastSnapshot.data );
+			// eslint-disable-next-line react/no-did-update-set-state
+			this.setState( {
+				diffData,
+				dataHistory: updatedHistory,
+				lastUpdated: now,
+			} );
+		}
+	}
+
+	updateHistory( history, data ) {
+		// Timestamp the new data snapshot.
+		const newSnapshot = {
+			timestamp: moment(),
+			data: data,
+		};
+
+		// Filter out snapshots older than minutesLimit prop.
+		// This determines the baseline for the diff calculation.
+		const { minutesLimit } = this.props;
+		const filteredHistory = [ ...history, newSnapshot ].filter(
+			( snapshot ) => moment().diff( snapshot.timestamp, 'minutes' ) <= minutesLimit
+		);
+
+		return this.compactHistory( filteredHistory );
+	}
+
+	compactHistory( history ) {
+		const MAX_HISTORY_LENGTH = 35;
+
+		if ( history.length > MAX_HISTORY_LENGTH ) {
+			// Keep every other entry to keep memory usage low.
+			return history.filter( ( _, index ) => index % 2 === 0 );
```

Comment:
Does this mean the refresh period would double after the history length is greater than `35`?

Reply:
No, the refresh period is not changed. All this does is check the length of the array and if over `MAX_HISTORY_LENGTH` it drops every 2nd element. So the array is 36 before the loop, then it's 18 after the loop.

The idea here is to thin the array but to keep the data spread across the full window.